stream-chat-android-client/src/main/java/io/getstream/chat/android/client/plugin/Plugin.kt

```diff
@@ -68,6 +70,9 @@ public interface Plugin :
     ChannelMarkReadListener,
     CreateChannelListener {
 
+    public val interceptors: List<Interceptor>
```

Comment:
Should plugins have interceptors? Plugins should only trigger side effects after backend calls. Holding interceptors and error handlers looks like adding more responsibilities than the original one. It would be better to keep Plugins with one single responsibility.

Reply:
Why plugin should only trigger side effects "after" backend calls?
Interceptors and ErrorHandler are kinds of side effects that we need to apply as well

Reply:
Sorry, I meant after and before backend calls, not only after it. 

Although plugins can have many functionalities, it would be better to narrow them down to just 1 or fewer functionalities, so it becomes more well-defined. 

The other methods are all pure side effect functions, so none of them have return types, or at least the return type is `Result<Unit>`. Adding the interceptors in the signature of `Plugin`, adds now the possibility of changing the information before it goes to the backend. 

If I had to define the responsibility of a Plugin I would say that it is a class that changes the state of the SDK before and after backend calls. But right now we are adding the ability to `intercept and change information,` `handle errors` and `connect and disconnect with the SDK`. 

I like the idea of accessing all the information from `StatePlugin` and `OfflinePlugin` so it becomes easier to understand how the adding/removing `state` and `offline` impacts the SDK, but I believe we need another interface to be the holder of the information. Like: 

```
public interface AddOn {
    public val plugins: List<Plugin>
    public val interceptors: List<Interceptor>
    public val errorHandler: List<ErrorHandler>
    
    public fun onUserSet(user: User)

    public fun onUserDisconnected()
}
```

It wouldn't change much from the current approach, but it would leave the `Plugin` interface untouched.

Reply:
But there's a more important observation about `interceptors` and `errorHandlers`. Differently than`Plugin`s,  `ErrorHandler`s, and `Interceptor`s **can't guarantee commutative property** (that's why `ErrorHandler` has the `priority` property). So you can never guarantee that A->B->C == C->B->A. In their case, order matters.

Because of that, interceptors and error handlers should all be in the same pool, so they can be properly sorted if a use case needs that. Otherwise we would have separated pools which won't allow sorting: 
```
[PluginA(ErrorHandler1, ErrorHandler2, ErrorHandler3), PluginA(ErrorHandler4, ErrorHandler5, ErrorHandler6)]
```

In the above case, it would be impossible to put `ErrorHandler4` before `ErrorHandler1` because PluginA comes first.. 

By adding `Interceptor` and `ErrorHandler` inside `Plugin`, we would have to worry about the order in plugins too, which makes it a lot more complex.

Sorry about almost writing a book...

Reply:
I agree with you, bu it is out of scope of this PR.
Previously we had "Plugin Factories" that:
1-. Create Plugins
2-. Inject dependencies that were needed in our Plugins into our ChatClient
3-. Sync lifecycle or our SDK and plugins.

Now, our "Plugin Factories" only does one thing, "Create Plugins".
The Plugin Instance is the one that handle the other 2 things.
Maybe we can innclude a new `AddOn` interface, but we will need to create some implementation for that, maybe some factory that helps our customer to use it and change the way we inject it into our `ChatClient`.
Not a trivial change that would need investigation/definition of how we want to go on.